pkg/i2gw/providers/gce/gce_extensions.go

```diff
@@ -0,0 +1,225 @@
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+type serviceList []types.NamespacedName
+
+func getBackendConfigMapping(ctx context.Context, storage *storage) (beConfigToSvcs map[types.NamespacedName]serviceList, betaBEConfigToSvcs map[types.NamespacedName]serviceList) {
+	beConfigToSvcs = make(map[types.NamespacedName]serviceList)
+	betaBEConfigToSvcs = make(map[types.NamespacedName]serviceList)
+
+	for _, service := range storage.Services {
+		svc := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+		ctx = context.WithValue(ctx, serviceKey, service)
+
+		// Read BackendConfig based on v1 BackendConfigKey.
+		beConfigName, exists := getBackendConfigName(ctx, service, annotations.BackendConfigKey)
+		if exists {
+			beConfigKey := types.NamespacedName{Namespace: service.Namespace, Name: beConfigName}
+			beConfigToSvcs[beConfigKey] = append(beConfigToSvcs[beConfigKey], svc)
+			continue
+		}
+
+		// Read BackendConfig based on v1beta1 BackendConfigKey.
+		beConfigName, exists = getBackendConfigName(ctx, service, annotations.BetaBackendConfigKey)
+		if exists {
+			beConfigKey := types.NamespacedName{Namespace: service.Namespace, Name: beConfigName}
+			betaBEConfigToSvcs[beConfigKey] = append(betaBEConfigToSvcs[beConfigKey], svc)
+			continue
+		}
+	}
+	return beConfigToSvcs, betaBEConfigToSvcs
+}
+
+// Get names of the BackendConfig in the cluster based on the BackendConfig
+// annotation on k8s Services.
+func getBackendConfigName(ctx context.Context, service *apiv1.Service, backendConfigKey string) (string, bool) {
+	val, exists := getBackendConfigAnnotation(service, backendConfigKey)
+	if !exists {
+		return "", false
+	}
+
+	return parseBackendConfigName(ctx, val)
+}
+
+// Get the backend config annotation from the K8s service if it exists.
+func getBackendConfigAnnotation(service *apiv1.Service, backendConfigKey string) (string, bool) {
+	val, ok := service.Annotations[backendConfigKey]
+	if ok {
+		return val, ok
+	}
+	return "", false
+}
+
+// Parse the name of the BackendConfig based on the annotation only if the same
+// BackendConfig is used for all Service ports.
+func parseBackendConfigName(ctx context.Context, val string) (string, bool) {
+	service := ctx.Value(serviceKey).(*apiv1.Service)
+
+	configs := annotations.BackendConfigs{}
+	if err := json.Unmarshal([]byte(val), &configs); err != nil {
+		notify(notifications.ErrorNotification, "BackendConfig annotation is invalid json", service)
+		return "", false
+	}
+
+	if configs.Default == "" && len(configs.Ports) == 0 {
+		notify(notifications.ErrorNotification, "No BackendConfig's found in annotation", service)
+		return "", false
+	}
+
+	if len(configs.Ports) != 0 {
+		notify(notifications.ErrorNotification, "Only config with default is supported since HealthCheckPolicy is attached on the whole service", service)
+		return "", false
+	}
+	return configs.Default, true
+}
+
+func ensureGceServiceIR(ctx context.Context, storage *storage, ir *i2gw.IR) {
```

Comment:
this func could benefit from a better name and a comment

Reply:
ditto for `ensureGceServiceExtensions`

Reply:
yeah i prefer `build` or `translate` as prefixes for the methods e.g. `buildGceServiceIR`

Reply:
Done. Updated to `buildGceServiceIR` and `buildGceServiceExtensions`.